napari/components/chunk/_loader.py

```diff
@@ -1,52 +1,278 @@
+            The ChunkRequest if it was satisfied otherwise None.
+
+        Notes
+        -----
+        We return a ChunkRequest if we performed the load synchronously,
+        otherwise we return None is indicating that an asynchronous load
+        was intitiated. When the async load finishes the layer's
+        on_chunk_loaded() will be called from the GUI thread.
+        """
+        if self._load_synchronously(request):
+            return request
+
+        # Clear any pending requests for this specific data_id.
+        self._clear_pending(request.key.data_id)
+
+        # Sumbit the request asynchronously.
+        self._submit_async(request)
+
+    def _load_synchronously(self, request: ChunkRequest) -> bool:
+        """Return True if we loaded the request synchronously."""
+
+        # If async is enabled, loaded non-ndarray request async.
+        if not self.synchronous and not request.in_memory:
+            return False  # load async
+
+        # Load synchronously.
+        request.load_chunks()
+        return True
+
+    def _submit_async(self, request: ChunkRequest) -> None:
+        """Initiate an asynchronous load of the given request.
+
+        Parameters
+        ----------
+        request : ChunkRequest
+            Contains the arrays to load.
         """
+        LOGGER.debug("ChunkLoader._load_async: %s", request.key)
+
+        # Submit the future, have it call ChunkLoader._done when done.
+        future = self.executor.submit(_chunk_loader_worker, request)
+        future.add_done_callback(self._done)
+
+        # Store the future in case we need to cancel it.
+        self.futures.setdefault(request.key.data_id, []).append(future)
+
+    def _clear_pending(self, data_id: int) -> None:
+        """Clear any pending requests for this data_id.
+
+        Parameters
+        ----------
+        data_id : int
+            Clear all requests associated with this data_id.
+        """
+        LOGGER.debug("ChunkLoader._clear_pending %d", data_id)
+
+        # Get list of futures we submitted to the pool.
+        future_list = self.futures.setdefault(data_id, [])
 
-        request.load_chunks()  # Load all chunks synchronously.
-        return request
+        # Try to cancel all futures in the list, but cancel() will return
+        # False if the task already started running.
+        num_before = len(future_list)
+        future_list[:] = [x for x in future_list if x.cancel()]
+        num_after = len(future_list)
+        num_cleared = num_before - num_after
+
+        # Delete the list entirely if empty
+        if num_after == 0:
+            del self.futures[data_id]
+
+        # Log what we did.
+        if num_before == 0:
+            LOGGER.debug("ChunkLoader.clear_pending: empty")
+        else:
+            LOGGER.debug(
+                "ChunkLoader.clear_pending: %d of %d cleared -> %d remain",
+                num_cleared,
+                num_before,
+                num_after,
```

Comment:
```suggestion
                f'ChunkLoader.clear_pending: {num_cleared} of {num_before} '
                f'cleared -> {num_after} remain'
```

Not positive this is better, but it feels more like modern Python. The C-style % formatting is increasingly rare.

Reply:
With f-strings it will format all the strings, then call into the logging module, then perhaps 99.999% of the time it will discover the log-level is such that the logging is disabled, so the formatting was wasted effort.

This is why the logging module uses the older style of string formatting. Pylint wil flag this as an error or warning, which is the only reason I know about this guideline, from projects in the past that used pylint. Logging performance when logging is disabled is important, since that's the most common case, it will mostly be disabled.

See also:
https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/

There is potentially a way to do it in the "Use of alternative formatting styles" section of the logging cookbook, but it involves creating a new custom class for every log message:
https://docs.python.org/3/howto/logging-cookbook.html

I'd like to switch the async logging to JSON format which would wipe away this issue as well, although it might create other performance issues. Ultimately once in a while it would be good to time things with all the log messages totally commented out, and see if it's slowing things down even when disabled.

Reply:
Fascinating! Thank you for the explanation. 👍 Perhaps put those two links in a comment above the first usage? That way I won't be tempted to "fix" it in 6 months when I come across this code and have forgotten this conversation. 😅

Reply:
I added a comment about the string formatting.